api/src/auth/drivers/oauth2.ts

```diff
@@ -136,15 +138,31 @@ export class OAuth2AuthDriver extends LocalAuthDriver {
 			throw new InvalidCredentialsException();
 		}
 
+		const userPayload = {
+			provider,
+			first_name: userInfo[this.config.firstNameKey],
+			last_name: userInfo[this.config.lastNameKey],
+			email: email,
+			external_identifier: identifier,
+			role: this.config.defaultRoleId,
+			auth_data: tokenSet.refresh_token && JSON.stringify({ refreshToken: tokenSet.refresh_token }),
+		};
+
 		const userId = await this.fetchUserId(identifier);
 
 		if (userId) {
-			// Update user refreshToken if provided
-			if (tokenSet.refresh_token) {
-				await this.usersService.updateOne(userId, {
-					auth_data: JSON.stringify({ refreshToken: tokenSet.refresh_token }),
-				});
-			}
+			// Run hook so the end user has the chance to augment the
+			// user that is about to be updated
+			const updatedUserPayload = await emitter.emitFilter(
+				`auth.oauth2.update`,
+				userPayload,
+				{ identifier, provider: this.config.provider, accessToken: tokenSet.access_token, userInfo },
+				{ database: getDatabase(), schema: this.schema, accountability: null }
+			);
+
+			// Update user to update refresh_token and other properties that might have changed
+			await this.usersService.updateOne(userId, updatedUserPayload);
+
```

Comment:
So I foresee an issue. If the user has made changes manually, they will get overwritten the next time they log in. A simple solution would be to just get back the updated fields from the hook, instead of the whole user payload.

Only for reference, pls don't apply:
```suggestion
			// Run hook so the end user has the chance to augment the
			// user that is about to be updated
			const updatedUserPayload = await emitter.emitFilter(
				`auth.oauth2.update`,
				{},
				{ identifier, provider: this.config.provider, accessToken: tokenSet.access_token, userInfo },
				{ database: getDatabase(), schema: this.schema, accountability: null }
			);

			// Update user to update refresh_token and other properties that might have changed
			await this.usersService.updateOne(userId, {
			    auth_data: tokenSet.refresh_token && JSON.stringify({ refreshToken: tokenSet.refresh_token }),
			    ...updatedUserPayload,
			});

```

Reply:
@aidenfoxx thank you for the review and the Input. Yes that is indeed the case, they would get overwritten and not passing the user as the payload can prevent that.

Technically I can adapt it to this approach and our use case would still work. It would be a bit more effort as the implementation of the hook would need to extract the necessary data from the userInfo based on the config and enrich the user object.

But before I start, I'm wondering if the case, that a user changes their firstname, lastname or email within directus to something that is different to the value in the Oauth/OIDC Provider is common.

Our use case for automatic updates is usually when people get married, they change their name and don't manage their accounts in the different systems but only in the Oauth/OIDC Provider.

I'm trying to figure out what's better in most cases or what would users expect to happen intuitively. Update or No-Update. Could you explain in which use cases the non-update approach makes more sense?

Thanks for your help!

Reply:
It's a fair point. From my view it's unintuative to let users update their profile, only for it to be overridden the next time they log in. This behavior would also be different to the LDAP and SAML flows, which also do not sync user data on subsequent logins.

To my mind it would be best to keep this PR in line with how auth works currently, and make a decision on how to better improve the sync experience for all flows in another ticket (e.g. should SSO flows allow users to update their personal info).

Thoughts @rijkvanzanten @paescuj?

Reply:
While I understand the need for such updates, I very much agree with @aidenfoxx to discuss (and possibly adapt)  this separately. I think we would run into too many unexpected problems, right now.

Reply:
Perhaps you could support both by doing

```js
const updatedUserPayload = await emitter.emitFilter(
  `auth.oauth2.update`,
  {},
  { userPayload, identifier, provider: this.config.provider, accessToken: tokenSet.access_token, userInfo },
  { database: getDatabase(), schema: this.schema, accountability: null }
);
```

Then you could either do:
```js
// only updates the role
export default ({ filter, action }) => {
  filter(`auth.oauth2.update`, ({ accessToken }) => ({ 
    role: accessToken.directus_role
  }));
};
```
or 
```js
// updates name, role, etc.
export default ({ filter, action }) => {
  filter(`auth.oauth2.update`, ({ accessToken, userPayload }) => ({ 
    ...userPayload,
    role: accessToken.directus_role
  }));
};
```

Reply:
Thank you for your inputs @aidenfoxx and @paescuj. I agree if the other drivers don't update the user after initial creation it makes no sense to do it here.

I will update my PR accordingly.

Thank you @beasteers, that's a good idea to make it easier for consumers of the hook that want the auto update functionality. I'll use that approach, unless @aidenfoxx or @paescuj see a reason why I should not pass it along separately.

Reply:
@aidenfoxx @paescuj Changes are in :)

Reply:
Thanks @nicam and @beasteers! To me it seems like a reasonable solution 👍 But I'd like to have this confirmed by @rijkvanzanten / @aidenfoxx!

Reply:
I have no strong feelings either way. It might be slightly confusing having both `userPayload` and `userInfo`, but as long as it's documented it's probably not an issue,

Reply:
> It might be slightly confusing having both `userPayload` and `userInfo`

True 👍 

> but as long as it's documented it's probably not an issue,

Also true 😄

I just think it's important that we have a solution which also makes sense and is supportable in the long term.

Reply:
maybe `openidUserInfo` might be clearer?

Reply:
@aidenfoxx So do you need me to do anything else here? :)

Reply:
👨🏻👈🏻 poke

We're adhering to our [contributing guidelines](https://github.com/directus/directus/blob/main/contributing.md) a little stricter nowadays to make sure PRs actually get the chance to be reviewed and merged. As per those guidelines, this PR should technically be closed, but I recognize that there was a PR in flights already from before those policies went into effect, so I'll make an exception. I'll try to find some time to slot this in in the coming weeks, but can't promise any timelines 🙂